extension/src/background/messageListener/popupMessageListener.ts

```diff
@@ -1301,6 +1398,192 @@ export const popupMessageListener = (request: Request, sessionStore: Store) => {
     return { mnemonicPhrase: migratedMnemonicPhrase };
   };
 
+  const migrateAccounts = async () => {
+    const { balancesToMigrate, isMergeSelected, recommendedFee } = request;
+
+    const migratedMnemonicPhrase = migratedMnemonicPhraseSelector(
+      sessionStore.getState(),
+    );
+    const password = passwordSelector(sessionStore.getState());
+    if (!password || !migratedMnemonicPhrase)
+      return { error: "Authentication error" };
+
+    const newWallet = fromMnemonic(migratedMnemonicPhrase);
+    const keyIdList: string = await getKeyIdList();
+    const migrationErrors: {
+      source: string;
+      destination: string;
+      asset?: string;
+      error: any;
+    }[] = [];
+    const fee = xlmToStroop(recommendedFee).toFixed();
+    const server = stellarSdkServer(NETWORK_URLS.TESTNET);
+
+    /*
+      For each migratable balance, we'll go through the following steps:
+      1. We create a new keypair that will be the destination account
+      2. We send the minimum amount of XLM needed to create the destination acct and also provide
+        enough funds to create necessary trustlines
+      3. Replace the old source account with the destination account in redux and in local storage. 
+        When the user refreshes the app, they will already be logged into their new accounts.
+      4. Migrate the trustlines from the source account to destination
+      5. Start an account session with the destination account so the user can start signing tx's with their newly migrated account 
+    */
+
+    for (let i = 0; i < balancesToMigrate.length; i += 1) {
+      const {
+        publicKey,
+        xlmBalance,
+        minBalance,
+        trustlineBalances,
+        keyIdIndex,
+      } = balancesToMigrate[i];
+
+      const keyID = keyIdList[keyIdIndex];
+
+      // eslint-disable-next-line no-await-in-loop
+      const store = await _unlockKeystore({ password, keyID });
+
+      // eslint-disable-next-line no-await-in-loop
+      const sourceAccount = await server.loadAccount(publicKey);
+
+      // create a new keystore and migrate while replacing the keyId in the list
+      const newKeyPair = {
+        publicKey: newWallet.getPublicKey(keyIdIndex),
+        privateKey: newWallet.getSecret(keyIdIndex),
+      };
+
+      // eslint-disable-next-line no-await-in-loop
+      const transaction = await new TransactionBuilder(sourceAccount, {
+        fee,
+        networkPassphrase: Networks.TESTNET,
+      });
+
+      if (!xlmBalance) {
```

Comment:
Should this condition trigger a user facing error so users know why a balance cannot be migrated?

Reply:
Yup, I'll work this into the UI